File: src/domain/BankAccount.js
'use strict';

/**
 * BankAccount Aggregate Root
 * Reconstructed by replaying events (or from a snapshot + subsequent events).
 */

const ACCOUNT_STATUS = {
  OPEN: 'OPEN',
  CLOSED: 'CLOSED',
};

class BankAccount {
  constructor() {
    this.accountId = null;
    this.ownerName = null;
    this.balance = 0;
    this.currency = 'USD';
    this.status = null;
    this.version = 0;       // last applied event_number
    this.changes = [];       // uncommitted events
  }

  // ─── Factory / Reconstitution ─────────────────────────────────────────────

  static createNew({ accountId, ownerName, initialBalance = 0, currency = 'USD' }) {
    const account = new BankAccount();
    const event = {
      eventType: 'AccountCreated',
      data: { accountId, ownerName, initialBalance, currency },
    };
    account._applyAndRecord(event);
    return account;
  }

  static fromEvents(events) {
    const account = new BankAccount();
    for (const event of events) {
      account._apply(event.event_type, event.event_data, event.event_number);
    }
    return account;
  }

  static fromSnapshot(snapshotData, subsequentEvents = []) {
    const account = new BankAccount();
    account.accountId = snapshotData.accountId;
    account.ownerName = snapshotData.ownerName;
    account.balance = parseFloat(snapshotData.balance);
    account.currency = snapshotData.currency;
    account.status = snapshotData.status;
    account.version = snapshotData.version;

    for (const event of subsequentEvents) {
      account._apply(event.event_type, event.event_data, event.event_number);
    }
    return account;
  }

  // ─── Commands ─────────────────────────────────────────────────────────────

  deposit({ amount, description, transactionId }) {
    if (this.status !== ACCOUNT_STATUS.OPEN) {
      throw { code: 'ACCOUNT_CLOSED', message: 'Account is closed.' };
    }
    if (!amount || amount <= 0) {
      throw { code: 'INVALID_AMOUNT', message: 'Deposit amount must be positive.' };
    }
    this._applyAndRecord({
      eventType: 'MoneyDeposited',
      data: { accountId: this.accountId, amount, description, transactionId },
    });
  }

  withdraw({ amount, description, transactionId }) {
    if (this.status !== ACCOUNT_STATUS.OPEN) {
      throw { code: 'ACCOUNT_CLOSED', message: 'Account is closed.' };
    }
    if (!amount || amount <= 0) {
      throw { code: 'INVALID_AMOUNT', message: 'Withdrawal amount must be positive.' };
    }
    if (this.balance < amount) {
      throw { code: 'INSUFFICIENT_FUNDS', message: 'Insufficient funds.' };
    }
    this._applyAndRecord({
      eventType: 'MoneyWithdrawn',
      data: { accountId: this.accountId, amount, description, transactionId },
    });
  }

  close({ reason }) {
    if (this.status !== ACCOUNT_STATUS.OPEN) {
      throw { code: 'ACCOUNT_CLOSED', message: 'Account is already closed.' };
    }
    if (this.balance !== 0) {
      throw { code: 'NON_ZERO_BALANCE', message: 'Account balance must be zero before closing.' };
    }
    this._applyAndRecord({
      eventType: 'AccountClosed',
      data: { accountId: this.accountId, reason },
    });
  }

  // ─── Internal Apply Logic ─────────────────────────────────────────────────

  _applyAndRecord(event) {
    this._apply(event.eventType, event.data, this.version + 1);
    this.changes.push(event);
  }

  _apply(eventType, data, eventNumber) {
    switch (eventType) {
      case 'AccountCreated':
        this.accountId = data.accountId;
        this.ownerName = data.ownerName;
        this.balance = parseFloat(data.initialBalance) || 0;
        this.currency = data.currency || 'USD';
        this.status = ACCOUNT_STATUS.OPEN;
        break;

      case 'MoneyDeposited':
        this.balance = parseFloat((this.balance + parseFloat(data.amount)).toFixed(4));
        break;

      case 'MoneyWithdrawn':
        this.balance = parseFloat((this.balance - parseFloat(data.amount)).toFixed(4));
        break;

      case 'AccountClosed':
        this.status = ACCOUNT_STATUS.CLOSED;
        break;

      default:
        break;
    }
    this.version = eventNumber;
  }

  // ─── Helpers ──────────────────────────────────────────────────────────────

  toSnapshot() {
    return {
      accountId: this.accountId,
      ownerName: this.ownerName,
      balance: this.balance,
      currency: this.currency,
      status: this.status,
      version: this.version,
    };
  }

  flushChanges() {
    const changes = [...this.changes];
    this.changes = [];
    return changes;
  }
}

module.exports = { BankAccount, ACCOUNT_STATUS };

File: src/main/java/com/bank/domain/BankAccount.java
package com.bank.domain;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * BankAccount Aggregate Root.
 * <p>
 * Reconstructed by replaying events (event sourcing) or from a snapshot + tail events.
 * All state mutations happen exclusively through event application — commands
 * only validate invariants and then produce events via {@link #applyAndRecord}.
 * </p>
 */
public class BankAccount {

    private String accountId;
    private String ownerName;
    private BigDecimal balance;
    private String currency;
    private AccountStatus status;
    private int version;                     // last applied event_number
    private final List<DomainEvent> changes; // uncommitted events

    // ─── Constructor ──────────────────────────────────────────────────────────

    public BankAccount() {
        this.balance = BigDecimal.ZERO;
        this.version = 0;
        this.changes = new ArrayList<>();
    }

    // ─── Factory: Create Brand-New Account ────────────────────────────────────

    public static BankAccount createNew(String accountId, String ownerName,
                                        BigDecimal initialBalance, String currency) {
        BankAccount account = new BankAccount();
        Map<String, Object> data = new HashMap<>();
        data.put("accountId", accountId);
        data.put("ownerName", ownerName);
        data.put("initialBalance", initialBalance);
        data.put("currency", currency != null ? currency : "USD");

        account.applyAndRecord(new DomainEvent("AccountCreated", data));
        return account;
    }

    // ─── Factory: Reconstitute from Full Event Stream ─────────────────────────

    @SuppressWarnings("unchecked")
    public static BankAccount fromEvents(List<Map<String, Object>> eventRows) {
        BankAccount account = new BankAccount();
        for (Map<String, Object> row : eventRows) {
            String eventType = (String) row.get("event_type");
            Object rawData = row.get("event_data");
            Map<String, Object> eventData = (rawData instanceof Map)
                    ? (Map<String, Object>) rawData
                    : new HashMap<>();
            int eventNumber = ((Number) row.get("event_number")).intValue();
            account.apply(eventType, eventData, eventNumber);
        }
        return account;
    }

    // ─── Factory: Reconstitute from Snapshot + Subsequent Events ──────────────

    @SuppressWarnings("unchecked")
    public static BankAccount fromSnapshot(Map<String, Object> snapshotData,
                                           List<Map<String, Object>> subsequentEvents) {
        BankAccount account = new BankAccount();
        account.accountId = (String) snapshotData.get("accountId");
        account.ownerName = (String) snapshotData.get("ownerName");
        account.balance = toBigDecimal(snapshotData.get("balance"));
        account.currency = (String) snapshotData.get("currency");
        account.status = AccountStatus.valueOf((String) snapshotData.get("status"));
        account.version = ((Number) snapshotData.get("version")).intValue();

        for (Map<String, Object> row : subsequentEvents) {
            String eventType = (String) row.get("event_type");
            Object rawData = row.get("event_data");
            Map<String, Object> eventData = (rawData instanceof Map)
                    ? (Map<String, Object>) rawData
                    : new HashMap<>();
            int eventNumber = ((Number) row.get("event_number")).intValue();
            account.apply(eventType, eventData, eventNumber);
        }
        return account;
    }

    // ─── Commands ─────────────────────────────────────────────────────────────

    public void deposit(BigDecimal amount, String description, String transactionId) {
        if (status != AccountStatus.OPEN) {
            throw new DomainException("ACCOUNT_CLOSED", "Account is closed.");
        }
        if (amount == null || amount.compareTo(BigDecimal.ZERO) <= 0) {
            throw new DomainException("INVALID_AMOUNT", "Deposit amount must be positive.");
        }

        Map<String, Object> data = new HashMap<>();
        data.put("accountId", accountId);
        data.put("amount", amount);
        data.put("description", description);
        data.put("transactionId", transactionId);
        applyAndRecord(new DomainEvent("MoneyDeposited", data));
    }

    public void withdraw(BigDecimal amount, String description, String transactionId) {
        if (status != AccountStatus.OPEN) {
            throw new DomainException("ACCOUNT_CLOSED", "Account is closed.");
        }
        if (amount == null || amount.compareTo(BigDecimal.ZERO) <= 0) {
            throw new DomainException("INVALID_AMOUNT", "Withdrawal amount must be positive.");
        }
        if (balance.compareTo(amount) < 0) {
            throw new DomainException("INSUFFICIENT_FUNDS", "Insufficient funds.");
        }

        Map<String, Object> data = new HashMap<>();
        data.put("accountId", accountId);
        data.put("amount", amount);
        data.put("description", description);
        data.put("transactionId", transactionId);
        applyAndRecord(new DomainEvent("MoneyWithdrawn", data));
    }

    public void close(String reason) {
        if (status != AccountStatus.OPEN) {
            throw new DomainException("ACCOUNT_CLOSED", "Account is already closed.");
        }
        if (balance.compareTo(BigDecimal.ZERO) != 0) {
            throw new DomainException("NON_ZERO_BALANCE", "Account balance must be zero before closing.");
        }

        Map<String, Object> data = new HashMap<>();
        data.put("accountId", accountId);
        data.put("reason", reason);
        applyAndRecord(new DomainEvent("AccountClosed", data));
    }

    // ─── Internal Apply Logic ─────────────────────────────────────────────────

    private void applyAndRecord(DomainEvent event) {
        apply(event.eventType(), event.data(), version + 1);
        changes.add(event);
    }

    private void apply(String eventType, Map<String, Object> data, int eventNumber) {
        switch (eventType) {
            case "AccountCreated" -> {
                this.accountId = (String) data.get("accountId");
                this.ownerName = (String) data.get("ownerName");
                this.balance = toBigDecimal(data.get("initialBalance"));
                this.currency = data.get("currency") != null ? (String) data.get("currency") : "USD";
                this.status = AccountStatus.OPEN;
            }
            case "MoneyDeposited" -> {
                BigDecimal amount = toBigDecimal(data.get("amount"));
                this.balance = this.balance.add(amount).setScale(4, RoundingMode.HALF_UP);
            }
            case "MoneyWithdrawn" -> {
                BigDecimal amount = toBigDecimal(data.get("amount"));
                this.balance = this.balance.subtract(amount).setScale(4, RoundingMode.HALF_UP);
            }
            case "AccountClosed" -> {
                this.status = AccountStatus.CLOSED;
            }
            default -> { /* ignore unknown events */ }
        }
        this.version = eventNumber;
    }

    // ─── Snapshot Helpers ─────────────────────────────────────────────────────

    public Map<String, Object> toSnapshot() {
        Map<String, Object> snap = new HashMap<>();
        snap.put("accountId", accountId);
        snap.put("ownerName", ownerName);
        snap.put("balance", balance);
        snap.put("currency", currency);
        snap.put("status", status.name());
        snap.put("version", version);
        return snap;
    }

    public List<DomainEvent> flushChanges() {
        List<DomainEvent> copy = new ArrayList<>(changes);
        changes.clear();
        return copy;
    }

    // ─── Getters ──────────────────────────────────────────────────────────────

    public String getAccountId()     { return accountId; }
    public String getOwnerName()     { return ownerName; }
    public BigDecimal getBalance()   { return balance; }
    public String getCurrency()      { return currency; }
    public AccountStatus getStatus() { return status; }
    public int getVersion()          { return version; }

    // ─── Utility ──────────────────────────────────────────────────────────────

    private static BigDecimal toBigDecimal(Object value) {
        if (value == null) return BigDecimal.ZERO;
        if (value instanceof BigDecimal bd) return bd;
        if (value instanceof Number n) return BigDecimal.valueOf(n.doubleValue());
        return new BigDecimal(value.toString());
    }
}
